pkg/controllermanager/controller/shoot/shoot_hibernation_control.go

```diff
@@ -56,85 +60,144 @@ func (c *Controller) shootHibernationUpdate(oldObj, newObj interface{}) {
 		newSchedule = getShootHibernationSchedules(newShoot)
 	)
 
-	if !reflect.DeepEqual(oldSchedule, newSchedule) {
+	if !reflect.DeepEqual(oldSchedule, newSchedule) && len(newSchedule) > 0 {
 		key, err := cache.MetaNamespaceKeyFunc(newObj)
 		if err != nil {
 			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", newObj, err)
 			return
 		}
-		c.shootHibernationQueue.Add(key)
+		parsedSchedules, err := parseHibernationSchedules(newSchedule)
+		if err != nil {
+			gardenlogger.Logger.Infof("Could not parse hibernation schedules for shoot %s: %v", client.ObjectKeyFromObject(newShoot), err)
+			return
+		}
+		requeueAfter := nextHibernationTimeDuration(parsedSchedules, TimeNow())
+		c.shootHibernationQueue.AddAfter(key, requeueAfter)
 	}
 }
 
-func (c *Controller) shootHibernationDelete(obj interface{}) {
-	shoot, ok := obj.(*gardencorev1beta1.Shoot)
-	if !ok {
-		return
+// ControllerName is the name of the controller.
+const ControllerName = "hibernation"
+
+// TimeNow returns the current time. Exposed for testing.
+var TimeNow = time.Now
+
+// ParsedHibernationSchedule holds the loaded location, parsed cron schedule and information whether
+// the cluster should be hibernated or woken up.
+type ParsedHibernationSchedule struct {
+	Location *time.Location
+	Schedule cron.Schedule
+	Enabled  bool
+}
+
+// Next returns the next activation time in UTC, later than the given time.
```

Comment:
Can you make this more clear? What is "activation" time? What is "later than the given time"?